File: functional/faster/faster.go
package main

import (
    "fmt"
    "strconv"
)

type Fs func(int) string

func slow(x int) string { return "slow-" + strconv.Itoa(x) }
func heavy(x int) string { return "heavy-" + strconv.Itoa(x) }

func makefaster(f Fs) Fs {
    cache := make(map[int]string)
    return func(x int) string {
        if cache[x] == "" {
            var a = f(x)
            cache[x] = a
            return "added to cache : " + a
        }
        return "from cache : " + cache[x]
    }
}

func main() {
    n := 42

    fmt.Println(slow(n))
    f1 := makefaster(slow)
    f2 := makefaster(heavy)

    fmt.Println(f1(n))
    fmt.Println(f1(n))
    fmt.Println(f1(23))

    fmt.Println(f2(n))
    fmt.Println(f2(n))
    fmt.Println(f2(23))
}

File: functional/faster/faster.js
var slow = function (x) {
   return "slow-" + x ;
};

var heavy = function (x) {
   return "slower-" + x;
};

function makefast(f) {
    var cache = {}; 
    return function (x) {
        if (cache[x] == undefined) {
            var a = f(x);
            cache[x] = a;
            return "write to cache : " + a;
        } else {
            return "from cache : " + cache[x];
        } 
    } 
}

var f1 = makefast(slow); 
var f2 = makefast(heavy); 

console.log(f1(100));
console.log(f2(100));
console.log(f1(100));
console.log(f2(100));

console.log(f1(200));
console.log(f2(200));

/* When it's written to cache it says "writing - "
 * When it's from cache it just writes it.
 */



